pkg/controller/oneagent/oneagent_controller_test.go

```diff
@@ -2,27 +2,99 @@ package oneagent
 
 import (
 	"context"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"os"
 	"testing"
 
 	dynatracev1alpha1 "github.com/Dynatrace/dynatrace-oneagent-operator/pkg/apis/dynatrace/v1alpha1"
+	dtclient "github.com/Dynatrace/dynatrace-oneagent-operator/pkg/dynatrace-client"
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/kubernetes/scheme"
+	restclient "k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-func TestReconcileOneAgent_ReconcileOnEmptyEnvironment(t *testing.T) {
-	var (
-		name      = "dynatrace-oneagent"
-		namespace = "dynatrace"
-	)
+func mockBuildDynatraceClient(instance *dynatracev1alpha1.OneAgent) (dtclient.Client, error) {
+
+	commHosts := []dtclient.CommunicationHost{
+		dtclient.CommunicationHost{
+			Protocol: "https",
+			Host:     "https://endpoint1.dev.ruxitlabs.com/communication",
+			Port:     443,
+		},
+		dtclient.CommunicationHost{
+			Protocol: "https",
+			Host:     "https://endpoint2.dev.ruxitlabs.com/communication",
+			Port:     443,
+		},
+	}
+
+	dtc := new(MyDynatraceClient)
+	dtc.On("GetVersionForIp", "127.0.0.1").Return("1.2.3", nil)
+	dtc.On("GetCommunicationHosts").Return(commHosts, nil)
+	dtc.On("GetAPIURLHost").Return(dtclient.CommunicationHost{
+		Protocol: "https",
+		Host:     testAPIUrl,
+		Port:     443,
+	}, nil)
+
+	return dtc, nil
+}
+
+func initMockServer(t *testing.T) *httptest.Server {
+	list := &metav1.APIGroupList{
+		Groups: []metav1.APIGroup{
+			{
+				Name: "networking.istio.io",
+				Versions: []metav1.GroupVersionForDiscovery{
+					{GroupVersion: "v1alpha3", Version: "v1alpha3"},
+					{GroupVersion: "v1alpha3", Version: "v1alpha3"},
+				},
+			},
+		},
+	}
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		var resources interface{}
+		switch req.URL.Path {
+		case "/apis":
+			resources = list
+		default:
+			// t.Logf("unexpected request: %s", req.URL.Path)
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		output, err := json.Marshal(resources)
+		if err != nil {
+			t.Errorf("unexpected encoding error: %v", err)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		w.Write(output)
+	}))
+
+	return server
+}
+
+func setupReconciler(t *testing.T) (*ReconcileOneAgent, client.Client) {
+	os.Setenv(k8sutil.WatchNamespaceEnvVar, "dynatrace")
+
+	server := initMockServer(t)
+	defer server.Close()
 
 	oa := newOneAgentSpec()
-	oa.ApiUrl = "https://ENVIRONMENTID.live.dynatrace.com/api"
+	oa.ApiUrl = testAPIUrl
 	oa.Tokens = "token_test"
+	oa.EnableIstio = true
```

Comment:
Do we want to set `EnableIstio = true` for all tests? There is a test for basic reconciliation without checking for Istio behavior.